### PR TITLE
MzTab more switches to stream reader/writer

### PR DESCRIPTION
**Pull request by timosachsenberg**
## Description

replaces https://github.com/OpenMS/OpenMS/pull/6195

## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

### CODE REVIEW

File: src/tests/topp/ProteinQuantifier_9_output_mztab.mzTab [hunk 1]
```diff
@@ -50,14 +50,14 @@ PRH	accession	description	taxid	species	database	database_version	search_engine
...
-PEP	EEEK	Protein	1	null	null	null	1.0	null	1.0	null	null	300.0	null	2	300.0	null	30.0	null	null	30.0	null	null	30.0	null	null	EEEK
-PEP	GGG	Protein	1	null	null	null	1.0	null	null	1.0	null	400.0	null	2	400.0	null	4.0	null	null	4.0	null	null	null	null	null	GGG
+PEH	sequence	accession	unique	database	database_version	search_engine	best_search_engine_score[1]	search_engine_score[1]_ms_run[1]	search_engine_score[1]_ms_run[2]	search_engine_score[1]_ms_run[3]	modifications	retention_time	retention_time_window	charge	mass_to_charge	spectra_ref	peptide_abundance_study_variable[1]	peptide_abundance_stdev_study_variable[1]	peptide_abundance_std_error_study_variable[1]	peptide_abundance_study_variable[2]	peptide_abundance_stdev_study_variable[2]	peptide_abundance_std_error_study_variable[2]	peptide_abundance_study_variable[3]	peptide_abundance_stdev_study_variable[3]	peptide_abundance_std_error_study_variable[3]	opt_global_id_merge_index	opt_global_cv_MS:1000889_peptidoform_sequence
+PEP	AAAK	Protein	1	null	null	null	1.0	1.0	null	null	null	100.0	null	2	100.0	null	1000.0	null	null	null	null	null	1000.0	null	null	null	AAAK
```
Comment (jpfeuffer) [inline src/tests/topp/ProteinQuantifier_9_output_mztab.mzTab:54]: Do you know why this is added, although all the values are null?

File: src/tests/topp/MzTabExporter_7_output.mzTab [hunk 2]
```diff
@@ -89,8 +89,8 @@ PRH	accession	description	taxid	species	database	database_version	search_engine
...
-PEP	null	null	null	null	null	null	null	null	null	460.306214458680699	null	2	430.185982351020698	null	1.851918e06	null	null	21.41680908203125	576	16393779085856951517	30433	0.932299655076805	0.803239812093931	473	controllerType=0 controllerNumber=1 scan=822
+PEH	sequence	accession	unique	database	database_version	search_engine	best_search_engine_score[1]	search_engine_score[1]_ms_run[1]	modifications	retention_time	retention_time_window	charge	mass_to_charge	spectra_ref	peptide_abundance_study_variable[1]	peptide_abundance_stdev_study_variable[1]	peptide_abundance_std_error_study_variable[1]	opt_global_FWHM	opt_global_cf_id	opt_global_feature_id	opt_global_label	opt_global_score_correlation	opt_global_score_fit	opt_global_spectrum_index	opt_global_spectrum_native_id	opt_global_cv_MS:1000889_peptidoform_sequence
+PEP	null	null	null	null	null	null	null	null	null	460.306214458680699	null	2	430.185982351020698	null	1.851918e06	null	null	21.41680908203125	576	16393779085856951517	30433	0.932299655076805	0.803239812093931	473	controllerType=0 controllerNumber=1 scan=822	null
```
Comment (jpfeuffer) [inline src/tests/topp/MzTabExporter_7_output.mzTab:93]: Why is it null?

Comment (timosachsenberg) [reply src/tests/topp/MzTabExporter_7_output.mzTab:93]: no sequence (also null)

Comment (jpfeuffer) [reply src/tests/topp/MzTabExporter_7_output.mzTab:93]: ok yes, makes sense

File: src/tests/topp/ProteinQuantifier_9_output_mztab.mzTab (hunk 1 above)
Comment (timosachsenberg) [reply src/tests/topp/ProteinQuantifier_9_output_mztab.mzTab:54]: I will check but I think this is a meta value added to the PeptideHit or Id during merging

Comment (jpfeuffer) [reply src/tests/topp/ProteinQuantifier_9_output_mztab.mzTab:54]: yes, but if all are null, it should not appear as a column, right? Either all IDs have that MV or none.

Comment (timosachsenberg) [reply src/tests/topp/ProteinQuantifier_9_output_mztab.mzTab:54]: could be a bug. I think it is a meta value at the PeptideIdentification level
https://github.com/OpenMS/OpenMS/blob/develop/src/tests/class_tests/openms/data/ProteinQuantifier_input.consensusXML#L32
and then read from the peptide hits. maybe we don't expose the peptideidentification meta values at all?

Comment (jpfeuffer) [reply src/tests/topp/ProteinQuantifier_9_output_mztab.mzTab:54]: Yes I remember that or a similar limitation

**Review by jpfeuffer — APPROVED** [review]
Looks good but double-check the tests carefully before merging. Especially with some tests always failing right now.

**Comment by jpfeuffer** [comment]
I think it is a very worthwhile TODO. I can imagine many feature infos that one would like to write in the future. Maybe if it is not too complicated, you could add it now, while you are still "into" the code.

**Comment by timosachsenberg** [comment]
k I can give it a try

**Comment by timosachsenberg** [comment]
k I can try to add it. let's do this incrementally as this PR seems complete. I will open a new one

**Comment by timosachsenberg** [comment]
failing tests unrelated

**Pull request merged**